src/page_parser.py

```diff
@@ -0,0 +1,362 @@
+# %pip install pdfminer.six
+import pdfminer
+import pdfminer.high_level as pdf_hl
+import pdfminer.layout as pdf_layout
+import os
+
+DO_IMAGE = True
+DO_TEST = True
+if DO_IMAGE:
+    import PIL
+    from PIL import Image
+    from PIL import ImageDraw
+
+
+def mirror_bbox_vertical(bbox, height):
+    x0, y0, x1, y1 = bbox
+    return (x0, height - y1, x1, height - y0)
+
+
+def mix_col(c1, c2, alpha):
+    c1_a = c1[3] / 255
+    c2_a = c2[3] / 255
+    tot_a = min(c1_a + c2_a, 1)
+    c1_t = tuple([c1[i] * c1[3] / 255 for i in range(3)])
+    c2_t = tuple([c2[i] * c2[3] / 255 for i in range(3)])
+    c_t = tuple(
+        [(c1_t[i] * alpha + c2_t[i] * (1 - alpha)) for i in range(3)] + [tot_a * 255]
+    )
+    return tuple([int(c_t[i]) for i in range(4)])
+
+
+def recursive_dict_add(d1, d2):
+    for k, v in d2.items():
+        if k in d1:
+            if type(d1[k]) == dict:
+                recursive_dict_add(d1[k], v)
+            else:
+                d1[k] += v
+        else:
+            d1[k] = v
+
+
+class Page_Parser:
+    def __init__(self, page: pdf_layout.LTPage):
+        self.page: pdf_layout.LTPage = page
+        self.unpacked: bool = False
+        self.bbox: tuple = page.bbox
+        self.width: int = page.width
+        self.height: int = page.height
+        self.grid = [
```

Comment:
if you want make these variables private and possibly add getter and setter functions if needed